EIP-4747: Simplify EIP-161 (#4747)
ethereum/EIPs: EIPS/eip-4747.md

---
eip: 4747
title: Simplify EIP-161
description: Simplify EIP-161 and retroactively deprecate unused aspects of it
author: Peter Davies (@petertdavies)
discussions-to: https://ethereum-magicians.org/t/eip-4747-simplify-eip-161/8246
status: Draft
type: Standards Track
category: Core
created: 2022-02-02
requires: 161
---

## Abstract

> Simplify the definition of EIP-161, removing the requirement for implementors to support edge cases that are impossible on Mainnet.

Reviewer: At least the first occurrence must be a link:
```suggestion
Simplify the definition of [EIP-161](./eip-161.md), removing the requirement for implementors to support edge cases that are impossible on Mainnet.
```
```suggestion
Simplify the definition of EIP-161, removing the requirement for implementors to support edge cases that are impossible on Mainnet.
```

## Motivation

> EIP-161 is overly complex and has a number of edge cases that are poorly documented and tested. This EIP takes advantage of the complete removal of all remaining empty accounts in block 14,049,881 (tx `0xf955834bfa097458a9cf6b719705a443d32e7f43f20b9b0294098c205b4bcc3d`) to clarify it, and allows implementors to not implement various edge cases that never occurred and are not possible in the future.

Reviewer: This EIP uses different styles for block numbers: it is comma delimited here and in the rationale, while without delimiters elsewhere.

My impression is that most EIPs do not include commas in block numbers, but it is up to you.

In particular, this EIP permits implementors who do not wish to support historical blocks to not implement state clearing at all.

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> Retroactively replace EIP-161, starting from the beginning of Spurious Dragon, with the following rules:

Reviewer: I think a actual block number is more useful here:
```suggestion
Retroactively replace EIP-161, starting from block 1... (aka. the activation of Spurious Dragon), with the following rules:
```

This EIP formally undefines all interactions between `STATICCALL` and state clearing as it has never happened on Mainnet and cannot happen in future.

### "At the end of the transaction"

This only makes a difference if an account is deleted and later recreated in the same transaction. This never happens on Mainnet.

### Test Cases

All test cases involving empty accounts in the Ethereum execution layer test suite shall be removed unless they relate to the Spurious Dragon Hardfork. If a Spurious Dragon test relates involved deprecated edgecase the test must be removed or reworked.

### Other networks

Ropsten had empty accounts seeded at genesis. They appear to have been cleared early in Ropsten's history before the Byzantium Hardfork. Ropsten has never been checked for edgecases occurring. All other Ethereum testnets have had EIP-161 from genesis.

As a security precaution all empty accounts on Ethereum Classic have been cleared, but no checks for edgecases occurring have been done. Due to EIP-161's age the vast majority of EVM compatible networks have supported it from genesis.

## Security considerations

> This EIP is only equivalent to EIP-161 on Mainnet if the following facts and must not be finalised until they have been independently verified:

Reviewer 2: ```suggestion
This EIP is only equivalent to EIP-161 on Ethereum Mainnet if the following facts are true:
```
EIPs should be written in their final form, which is presumably *after* independent verification has occurred.  I recommend mentioning the need for independent verification in the discussions-to link.

1. No empty accounts are ever touched and then reinstated in the same transaction.

> 2. The transactions in the Appendix are the only state clearing transactions on Mainnet after the start of Byzantium.

Reviewer: Should include block number for Byzantium here, imho.
